man/opkg.1.in
.TH "@PACKAGE_NAME@" 1 "@CLEAN_DATE@" "@PACKAGE_STRING@" "User Commands"
.SH NAME
opkg \- command line utility to install, upgrade and uninstall opkg
software packages
.
.SH SYNOPSIS
.B \fBopkg\fP [\fIOPTIONS...\fP] \fIsub-command\fP [\fIARGUMENTS...\fP]
.
.SH DESCRIPTION
\fBopkg\fP is a simple utility to install, upgrade and uninstall opkg
software packages. Opkg is a lightweight package management system based
on Ipkg.
.
.SH "OPTIONS AND COMMANDS"
.SS SUB-COMMANDS : PACKAGE MANIPULATION
.TP
\fBupdate\fR
Update list of available packages
.TP
\fBupgrade\fR
Upgrade installed packages
.TP
\fBdist-upgrade\fR
Upgrade operation that allows package removal. Only available on the libsolv backend.
.TP
\fBinstall <\fIpackage(s)\fP>\fR
Install \fIpackage(s)\fP
.TP
\fBconfigure <\fIpackage(s)\fP>\fR
Configure unpacked \fIpackage(s)\fP
.TP
\fBremove <\fIpackages\fP|\fIglob\fP>\fR
Remove \fIpackage(s)\fP. \fIglob\fP works like a shell globbing pattern and
could be something like 'pkgname*' '*file*' or similar
.TP
\fBclean \fR
Clean internal cache
.TP
\fBflag <\fIflag\fP> <\fIpackages\fP>\fR
Flag \fIpackage(s)\fP. Available flags (one per invocation):
.TS
tab(@);
l.
hold
noprune
user
ok
installed
unpacked
.TE
.
.TP
\fBversion \fR
Print version
.SS INFORMATIONAL SUB-COMMANDS
.TP
\fBlist\fR
List available packages
.TP
\fBlist-installed\fR
List installed packages
.TP
\fBlist-upgradable\fR
List installed and upgradable packages
.TP
\fBlist-changed-conffiles\fR
List package configuration files which have been modified after installation
.TP
\fBfiles <\fIpackage\fP>\fR
List files belonging to \fIpackage\fP
.TP
\fBsearch <\fIfile\fP|\fIglob\fP>\fR
List package providing \fIfile\fP
.TP
\fBfind <\fIregexp\fP>\fR
List packages with names or description matching \fIregexp\fP
.TP
\fBinfo [\fIpackage\fP|\fIglob\fP]\fR
Display all info for selected packages
.TP
\fBstatus [\fIpackage\fP|\fIglob\fP]\fR
Display all statuses for selected packages
.TP
\fBdownload <\fIpackage\fP>\fR
Download \fIpackage\fP to current directory
.TP
\fBcompare-versions <\fIversion1\fP> <\fIoperator\fP> <\fIversion2\fP>\fR
compare versions using following operators :
.TS
tab(@);
c l.
<<@less than
<@less than or equal to
<=@less than or equal to
 =@equal to
>=@greater than or equal to
>@greater than or equal to
>>@greater than
.TE
.TP
\fBprint-architecture\fR
List installable package architectures
.TP
\fBdepends [\fI\-A\fP] <\fIpackage(s)\fP|\fIglob\fP>\fR
.TP
\fBwhatdepends [\fI\-A\fP] <\fIpackage(s)\fP|\fIglob\fP>\fR
.TP
\fBwhatdependsrec [\fI\-A\fP] <\fIpackage(s)\fP|\fIglob\fP>\fR
.TP
\fBwhatrecommends [\fI\-A\fP] <\fIpackage(s)\fP|\fIglob\fP>\fR
.TP
\fBwhatsuggests [\fI\-A\fP] <\fIpackage(s)\fP|\fIglob\fP>\fR
.TP
\fBwhatprovides [\fI\-A\fP] <\fIpackage(s)\fP|\fIglob\fP>\fR
.TP
\fBwhatconflicts [\fI\-A\fP] <\fIpackage(s)\fP|\fIglob\fP>\fR
.TP
\fBwhatreplaces [\fI\-A\fP] <\fIpackage(s)\fP|\fIglob\fP>\fR
.TP
\fBverify  <\fIpackage(s)\fP|\fIglob\fP>\fR
Verifies the integrity of <pkg>, or all packages if omitted by
comparing the md5sum of each file with the information stored
on the opkg metadata database.

.SS OPTIONS
.TP
\fB\-A\fR
Query all packages not just those installed
.
.TP
\fB\-V <\fIlevel\fP>, --verbosity <\fIlevel\fP>\fR
Set verbosity level to \fIlevel\fP. Verbosity levels :
.TS
tab(@);
l l.
0@errors only
1@normal messages (default)
2@informative messages
3@debug
4@debug level 2
.TE
.
.TP
\fB\-f <\fIconf_file\fP>, \fB\--conf <\fIconf_file\fP>\fR
Use \fIconf_file\fP as the opkg configuration file
.TP
\fB\--cache-dir <\fIdirectory\fP>\fR
Use a package cache
.TP
\fB\-t <\fIdirectory\fP>, \--tmp-dir <\fIdirectory\fP>\fR
Specify \fIdirectory\fP as temporary directory
.TP
\fB\-l <\fIdirectory\fP>, \--lists-dir <\fIdirectory\fP>\fR
Specify \fIdirectory\fP as package lists directory
.TP
\fB\-d <\fIdest_name\fP>, \fB\--dest <\fIdest_name\fP>\fR
Use \fIdest_name\fP as the the root directory for
package installation, removal, upgrading. \fIdest_name\fP should be a
defined dest name from the configuration file, (but can also be a
directory name in a pinch).
.TP
\fB\-o <\fIdirectory\fP>, \fB\--offline-root <\fIdirectory\fP>\fR
Use \fIdirectory\fP as the root directory for offline installation of
packages.
.TP
\fB\--add-dest <\fIname\fP>:<\fIpath\fP>\fR
Register \fIpath\fP as installation target \fIname\fP for use in
conjunction with \fB\--dest\fP
.TP
\fB\--add-arch <\fIarch\fP>:<\fIprio\fP>\fR
Register the package architecture \fIarch\fP with the numeric
priority \fIprio\fP. Lower priorities take precedence.
.TP
\fB\--add-exclude <\fIname\fP>\fR
Register package to be excluded from install
.TP
\fB\--add-ignore-recommends <\fIname\fP>\fR
Register package to be ignored as a recomendee
.TP
\fB\--prefer-arch-to-version\fR
Use the architecture priority package rather than the higher version
one if more than one candidate is found.
.TP
\fB\--combine\fR
Combine upgrade and install operations, this may be needed to resolve
dependency issues. Only available for the internal solver backend.
.TP
\fB\--fields <\fIfield1\fP>,<\fIfield2\fP>\fR
Limit display information to the specified fields plus
the package name. Valid for \fBinfo\fP and \fBstatus\fP.
.TP
\fB\--short-description
Display only the first line of the description.
.TP
\fB\--size\fR
Print package size when listing available packages
.SS FORCE OPTIONS
.TP
\fB\--force-depends \fR
Install/remove despite failed dependencies
.TP
\fB\--force-maintainer \fR
Overwrite preexisting config files
.TP
\fB\--force-reinstall \fR
Reinstall package(s)
.TP
\fB\--force-overwrite\fR
Overwrite files from other package(s)
.TP
\fB\--force-downgrade\fR
Allow opkg to downgrade packages
.TP
\fB\--force-space \fR
Disable free space checks
.TP
\fB\--force-postinstall \fR
Execute package postinstall scripts in offline installation mode
.TP
\fB\--force-remove \fR
Remove package even if prerm script fails
.TP
\fB\--force-checksum \fR
Don't fail on checksum mismatches
.TP
\fB\--noaction\fR
No action \- test only
.TP
\fB\--download-only\fR
No action \- download only
.TP
\fB\--nodeps\fR
Do not follow dependencies
.TP
\fB\--no-install-recommends\fR
Do not install any recommended packages
.TP
\fB\--force-removal-of-dependent-packages\fR
Remove package and all dependencies
.TP
\fB\--autoremove\fR
Remove packages that were installed automatically to satisfy dependencies
.TP
\fB\--host-cache-dir\fR
Don't place cache in offline root dir
.TP
\fB\--volatile-cache\fR
Use volatile cache. Volatile cache will be cleared on exit
.
.SH "REPORTING BUGS"
Report bugs to https://bugzilla.yoctoproject.org/buglist.cgi?product=opkg
.
.P
Copyright \(co Opkg development team.
.P
Copyright \(co Ipkg development team.
.


NEWS
opkg-0.4.3
==========

- Do not populate install status if package is not installed (issue 13760)

- Add glob support for --add-ignore-recommends and --add-exclude (issue 13759)

- Fix possible bad memory access in file_read_line_alloc

- Allow upgrades between packages with different names

- Remove extra slash on dest struct variables

- Prevent double free on opkg_conf_load

Libsolv:

- Show all solver problems and not only an educated guess

- Ignore missing recommends of installed packages (issue 13758)

opkg-0.4.2
==========

- Follow symlinks during install (issue 13574).

- Add support for empty payloads. Before, an empty payload would segfault opkg.

- Add new opkg verify option, to check system status.

- Add SPDX license identifiers to source files.

- Speed up opkg clean by avoiding loading feed/status information.

- Fix static build (issue 8355).

- Add missing pkg->tags initialization (issue 13380).

opkg-0.4.1
==========

- Improve support for Debian-based repos (issue 13185).

- Add target for testsuite installation.

- Add --add-ignore-recommends option to ignore specific recommendations.

- Check download file size when checking file hash, to insulate verification from hash collisions.

- Abort operations on download error.

- Move gpg homedir to /etc/opkg/gpg to decouple opkg from gpg internals. This
  is a breaking change if gpg is enabled and opkg is upgraded. You can use the
  conf option "option gpg dir /etc/opkg" to revert back to the old directory
  layout.

- Avoid window with overly permissive umask when creating directory hierarchy 

Libsolv:

- Make --no-install-recommends case libsolv flags consistent with the canonical case.

opkg-0.4.0
==========

- Remove support for long options with only one leading dash.

- Use strict matching, instead of globs, when removing maintaner scripts (issue
  12905).

- Speed up opkg installed files algorithm (issue 12648). This was a regression
  from 0.3.4.

- Add support for lz4 compression.

- Add --fields command line argument, to customize 'info' and 'status' output.

- Add --short-description, to limit the output of 'info' and 'status'.

- Error out if feeds only have sha256 sums, but --enable-sha256=false.

- Clean up command line help and man page.

Libsolv:

- Use arch when finding a package match. Before it was only using name and
  version, which lead to incorrect results if a feed had two packages with the
  name and version, but different arch.

- Make sure Essential package dependencies are not removed.

- Abort installation on first failure.

opkg-0.3.6
==========

- Optimize access to packagename.list file

- Add compress_list_files option to optimize disk space (issue 11912)

- Avoid duplicated entries on status file, due to stale package index (issue
  8351)

- Miscelaneous fixes to make opkg compatible on more paltforms.

Internal Solver:

- Enable upgrades of installed packages that Provide a virtual package already
  installed (issue 11926, regression from 0.3.4)

Libsolv:

- Add list-upgradable support (issue 8165)

- Run maintainer scripts in correct order when upgrading virtual packages
  (issue 11362)

- Make missing packages solvable non-version constrained

opkg-0.3.5
==========

- Handle conffiles that don't exist gracefully (show notice instead of error)

- Correctly handle symlink files collisions (issue 11033)

- Add verbose_status_file option to store all the package metadata in the status file

- Add support for user-defined control file fields

- Enhance PATH handling for intercepts

- Explicitly run scripts and intercepts with /bin/sh instead of sh

- Run maintainer scripts by default in the ATS

- Add support for installing packages with version constrains other than '='

Internal Solver:

- Do not process orphans in download-only case (issue 10243)

- Remove duplicate processing of provides and replacees

- Only install one provider for each virtual package (issue 10781)

- Skip dependency error message if --force-depends (issue 9060)

Libsolv:

- Allow removal of recommended packages, in case of conflict (issue 10777)

- During upgrades, also upgrade Recommends (issue 9939)

- Print solver errors to stderr (issue 10407)

- Add dist-upgrade support

opkg-0.3.4
==========

- Improve GPG signature check by making sure the signature is not only correct,
  but has an approved level of trust on the trusted database.

- Add support for xz and bzip2 compressed payloads.

- Handle 'opkg remove' and 'opkg install' (libsolv) correctly for packages in
  SS_HALF_INSTALLED state due to a previously failed installation (issues 10358
  & 9901).

- Remove redundant requirement to have check_signature enabled when
  check_pkg_signature is enabled (issue 10361)

- Add Installed-Size field to status file (issue 9906)

- Do not overwrite existing directory permissions on install (issue 9987)

- Add per-feed signature verification toggle (issue 10150)

- Add validation to opkg flag call (issue 8356)

- Add "find" command which searches package names and descriptions (e.g. opkg
  -i find "*autoconfig*")

- Add --force-checksum cmdline flag to be able to ignore mismatching md5sums
  (issue 8357)

- Expose lists_dir option as command line argument

- Extend 'opkg list' command to optionally display package size


src/opkg.c
/* vi: set expandtab sw=4 sts=4: */
/* opkg.c - the opkg package management system

   Florian Boor
   Copyright (C) 2003 kernel concepts

   Carl D. Worth
   Copyright 2001 University of Southern California

   SPDX-License-Identifier: GPL-2.0-or-later

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2, or (at
   your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   opkg command line frontend using libopkg
*/

#include "config.h"

#include <stdio.h>
#include <getopt.h>
#include <stdlib.h>

#include "opkg_conf.h"
#include "opkg_cmd.h"
#include "file_util.h"
#include "opkg_message.h"
#include "opkg_download.h"
#include "xfuncs.h"

enum {
    ARGS_OPT_FORCE_MAINTAINER = 129,
    ARGS_OPT_IGNORE_MAINTAINER,
    ARGS_OPT_FORCE_DEPENDS,
    ARGS_OPT_FORCE_OVERWRITE,
    ARGS_OPT_FORCE_DOWNGRADE,
    ARGS_OPT_FORCE_REINSTALL,
    ARGS_OPT_FORCE_REMOVAL_OF_DEPENDENT_PACKAGES,
    ARGS_OPT_FORCE_REMOVAL_OF_ESSENTIAL_PACKAGES,
    ARGS_OPT_FORCE_SPACE,
    ARGS_OPT_FORCE_POSTINSTALL,
    ARGS_OPT_FORCE_REMOVE,
    ARGS_OPT_FORCE_CHECKSUM,
    ARGS_OPT_PREFER_ARCH_TO_VERSION,
    ARGS_OPT_ADD_ARCH,
    ARGS_OPT_ADD_DEST,
    ARGS_OPT_SIZE,
    ARGS_OPT_ADD_EXCLUDE,
    ARGS_OPT_ADD_IGNORE_RECOMMENDS,
    ARGS_OPT_NOACTION,
    ARGS_OPT_DOWNLOAD_ONLY,
    ARGS_OPT_NODEPS,
    ARGS_OPT_AUTOREMOVE,
    ARGS_OPT_VOLATILE_CACHE,
    ARGS_OPT_COMBINE,
    ARGS_OPT_NO_INSTALL_RECOMMENDS,
    ARGS_OPT_CACHE_DIR,
    ARGS_OPT_HOST_CACHE_DIR,
    ARGS_OPT_SHORT_DESCRIPTION,
    ARGS_OPT_FIELDS_FILTER,
};

static struct option long_options[] = {
    {"query-all", 0, 0, 'A'},
    {"autoremove", 0, 0, ARGS_OPT_AUTOREMOVE},
    {"conf-file", 1, 0, 'f'},
    {"conf", 1, 0, 'f'},
    {"combine", 0, 0, ARGS_OPT_COMBINE},
    {"dest", 1, 0, 'd'},
    {"force-maintainer", 0, 0, ARGS_OPT_FORCE_MAINTAINER},
    {"force_maintainer", 0, 0, ARGS_OPT_FORCE_MAINTAINER},
    {"ignore-maintainer", 0, 0, ARGS_OPT_IGNORE_MAINTAINER},
    {"ignore_maintainer", 0, 0, ARGS_OPT_IGNORE_MAINTAINER},
    {"force-depends", 0, 0, ARGS_OPT_FORCE_DEPENDS},
    {"force_depends", 0, 0, ARGS_OPT_FORCE_DEPENDS},
    {"force-overwrite", 0, 0, ARGS_OPT_FORCE_OVERWRITE},
    {"force_overwrite", 0, 0, ARGS_OPT_FORCE_OVERWRITE},
    {"force_downgrade", 0, 0, ARGS_OPT_FORCE_DOWNGRADE},
    {"force-downgrade", 0, 0, ARGS_OPT_FORCE_DOWNGRADE},
    {"force-reinstall", 0, 0, ARGS_OPT_FORCE_REINSTALL},
    {"force_reinstall", 0, 0, ARGS_OPT_FORCE_REINSTALL},
    {"force-space", 0, 0, ARGS_OPT_FORCE_SPACE},
    {"force_space", 0, 0, ARGS_OPT_FORCE_SPACE},
    {"force-removal-of-dependent-packages", 0, 0,
     ARGS_OPT_FORCE_REMOVAL_OF_DEPENDENT_PACKAGES},
    {"force_removal_of_dependent_packages", 0, 0,
     ARGS_OPT_FORCE_REMOVAL_OF_DEPENDENT_PACKAGES},
    {"force-removal-of-essential-packages", 0, 0,
     ARGS_OPT_FORCE_REMOVAL_OF_ESSENTIAL_PACKAGES},
    {"force_removal_of_essential_packages", 0, 0,
     ARGS_OPT_FORCE_REMOVAL_OF_ESSENTIAL_PACKAGES},
    {"force-postinstall", 0, 0, ARGS_OPT_FORCE_POSTINSTALL},
    {"force_postinstall", 0, 0, ARGS_OPT_FORCE_POSTINSTALL},
    {"force-remove", 0, 0, ARGS_OPT_FORCE_REMOVE},
    {"force_remove", 0, 0, ARGS_OPT_FORCE_REMOVE},
    {"force-checksum", 0, 0, ARGS_OPT_FORCE_CHECKSUM},
    {"force_checksum", 0, 0, ARGS_OPT_FORCE_CHECKSUM},
    {"prefer-arch-to-version", 0, 0, ARGS_OPT_PREFER_ARCH_TO_VERSION},
    {"prefer_arch_to_version", 0, 0, ARGS_OPT_PREFER_ARCH_TO_VERSION},
    {"noaction", 0, 0, ARGS_OPT_NOACTION},
    {"download-only", 0, 0, ARGS_OPT_DOWNLOAD_ONLY},
    {"nodeps", 0, 0, ARGS_OPT_NODEPS},
    {"no-install-recommends", 0, 0, ARGS_OPT_NO_INSTALL_RECOMMENDS},
    {"offline-root", 1, 0, 'o'},
    {"add-arch", 1, 0, ARGS_OPT_ADD_ARCH},
    {"add-dest", 1, 0, ARGS_OPT_ADD_DEST},
    {"size", 0, 0, ARGS_OPT_SIZE},
    {"add-exclude", 1, 0, ARGS_OPT_ADD_EXCLUDE},
    {"add-ignore-recommends", 1, 0, ARGS_OPT_ADD_IGNORE_RECOMMENDS},
    {"tmp-dir", 1, 0, 't'},
    {"tmp_dir", 1, 0, 't'},
    {"lists-dir", 1, 0, 'l'},
    {"lists_dir", 1, 0, 'l'},
    {"cache-dir", 1, 0, ARGS_OPT_CACHE_DIR},
    {"host-cache-dir", 0, 0, ARGS_OPT_HOST_CACHE_DIR},
    {"volatile-cache", 0, 0, ARGS_OPT_VOLATILE_CACHE},
    {"short-description", 0, 0, ARGS_OPT_SHORT_DESCRIPTION},
    {"fields", 1, 0, ARGS_OPT_FIELDS_FILTER},
    {"verbosity", 2, 0, 'V'},
    {"version", 0, 0, 'v'},
    {0, 0, 0, 0}
};

static int args_parse(int argc, char *argv[])
{
    int c;
    int option_index = 0;
    int parse_err = 0;
    char *tuple, *targ;
    char *solver_version = NULL;

    while (1) {
        c = getopt_long(argc, argv, "Ad:f:no:p:l:t:vV::", long_options,
                             &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'A':
            opkg_config->query_all = 1;
            break;
        case 'd':
            opkg_config->dest_str = xstrdup(optarg);
            break;
        case 'f':
            opkg_config->conf_file = xstrdup(optarg);
            break;
        case 'o':
            opkg_config->offline_root = xstrdup(optarg);
            break;
        case 't':
            opkg_config->tmp_dir = xstrdup(optarg);
            break;
        case 'l':
            opkg_config->lists_dir = xstrdup(optarg);
            break;
        case 'v':
            solver_version = opkg_solver_version_alloc();
            if (solver_version) {
                printf("opkg version " VERSION " (%s)\n", solver_version);
                free(solver_version);
            } else {
                printf("opkg version " VERSION "\n");
            }
            exit(0);
        case 'V':
            opkg_config->verbosity = INFO;
            if (optarg != NULL)
                opkg_config->verbosity = atoi(optarg);
            break;
        case ARGS_OPT_AUTOREMOVE:
            opkg_config->autoremove = 1;
            break;
        case ARGS_OPT_FORCE_MAINTAINER:
            opkg_config->force_maintainer = 1;
            break;
        case ARGS_OPT_IGNORE_MAINTAINER:
            opkg_config->ignore_maintainer = 1;
            break;
        case ARGS_OPT_FORCE_DEPENDS:
            opkg_config->force_depends = 1;
            break;
        case ARGS_OPT_FORCE_OVERWRITE:
            opkg_config->force_overwrite = 1;
            break;
        case ARGS_OPT_FORCE_DOWNGRADE:
            opkg_config->force_downgrade = 1;
            break;
        case ARGS_OPT_FORCE_REINSTALL:
            opkg_config->force_reinstall = 1;
            break;
        case ARGS_OPT_FORCE_REMOVAL_OF_ESSENTIAL_PACKAGES:
            opkg_config->force_removal_of_essential_packages = 1;
            break;
        case ARGS_OPT_FORCE_REMOVAL_OF_DEPENDENT_PACKAGES:
            opkg_config->force_removal_of_dependent_packages = 1;
            break;
        case ARGS_OPT_FORCE_SPACE:
            opkg_config->force_space = 1;
            break;
        case ARGS_OPT_FORCE_POSTINSTALL:
            opkg_config->force_postinstall = 1;
            break;
        case ARGS_OPT_FORCE_REMOVE:
            opkg_config->force_remove = 1;
            break;
        case ARGS_OPT_FORCE_CHECKSUM:
            opkg_config->force_checksum = 1;
            break;
        case ARGS_OPT_PREFER_ARCH_TO_VERSION:
            opkg_config->prefer_arch_to_version = 1;
            break;
        case ARGS_OPT_NODEPS:
            opkg_config->nodeps = 1;
            break;
        case ARGS_OPT_ADD_ARCH:
        case ARGS_OPT_ADD_DEST:
            tuple = xstrdup(optarg);
            if ((targ = strchr(tuple, ':')) != NULL) {
                *targ++ = 0;
                if ((strlen(tuple) > 0) && (strlen(targ) > 0)) {
                    nv_pair_list_append((c == ARGS_OPT_ADD_ARCH) ? &opkg_config->arch_list : &opkg_config->tmp_dest_list,
                                        tuple, targ);
                }
            }
            free(tuple);
            break;
        case ARGS_OPT_ADD_EXCLUDE:
            str_list_append(&opkg_config->exclude_list, optarg);
            break;
        case ARGS_OPT_ADD_IGNORE_RECOMMENDS:
            str_list_append(&opkg_config->ignore_recommends_list, optarg);
            break;
        case ARGS_OPT_SIZE:
            opkg_config->size = 1;
            break;
        case ARGS_OPT_NOACTION:
            opkg_config->noaction = 1;
            break;
        case ARGS_OPT_DOWNLOAD_ONLY:
            opkg_config->download_only = 1;
            break;
        case ARGS_OPT_CACHE_DIR:
            opkg_config->cache_dir = xstrdup(optarg);
            break;
        case ARGS_OPT_HOST_CACHE_DIR:
            opkg_config->host_cache_dir = 1;
            break;
        case ARGS_OPT_VOLATILE_CACHE:
            opkg_config->volatile_cache = 1;
            break;
        case ARGS_OPT_NO_INSTALL_RECOMMENDS:
            opkg_config->no_install_recommends = 1;
            break;
        case ARGS_OPT_SHORT_DESCRIPTION:
           opkg_config->short_description = 1;
           break;
        case ARGS_OPT_FIELDS_FILTER:
            opkg_config->fields_filter = xstrdup(optarg);
            break;
        case ARGS_OPT_COMBINE:
            opkg_config->combine = 1;
            break;
        case ':':
            parse_err = -1;
            break;
        case '?':
            parse_err = -1;
            break;
        default:
            printf("Confusion: getopt_long returned %d\n", c);
        }
    }

    if (parse_err)
        return parse_err;
    else
        return optind;
}

static void usage()
{
    printf("usage: opkg [options...] sub-command [arguments...]\n");
    printf("where sub-command is one of:\n");

    printf("\nPackage Manipulation:\n");
    printf("\tupdate                          Update list of available packages\n");
    printf("\tupgrade                         Upgrade installed packages\n");
    printf("\tdist-upgrade                    Upgrade operation that allows package removal (libsolv backend)\n");
    printf("\tinstall <pkgs>                  Install package(s)\n");
    printf("\tconfigure <pkgs>                Configure unpacked package(s)\n");
    printf("\tremove <pkgs|glob>              Remove package(s)\n");
    printf("\tclean                           Clean internal cache\n");
    printf("\tflag <flag> <pkgs>              Flag package(s)\n");
    printf("\t <flag>=hold|noprune|user|ok|installed|unpacked (one per invocation)\n");

    printf("\nInformational Commands:\n");
    printf("\tlist                            List available packages\n");
    printf("\tlist-installed                  List installed packages\n");
    printf("\tlist-upgradable                 List installed and upgradable packages\n");
    printf("\tlist-changed-conffiles          List user modified configuration files\n");
    printf("\tfiles <pkg>                     List files belonging to <pkg>\n");
    printf("\tsearch <file|glob>              List package providing <file>\n");
    printf("\tfind <regexp>                   List packages with names or description matching <regexp>\n");
    printf("\tinfo [pkg|glob]                 Display all info for <pkg>\n");
    printf("\tstatus [pkg|glob]               Display all status for <pkg>\n");
    printf("\tdownload <pkg>                  Download <pkg> to current directory\n");
    printf("\tcompare-versions <v1> <op> <v2>\n");
    printf("\t                                compare versions using <= < > >= = << >>\n");
    printf("\tprint-architecture              List installable package architectures\n");
    printf("\tdepends [-A] [pkgname|glob]+\n");
    printf("\twhatdepends [-A] [pkgname|glob]+\n");
    printf("\twhatdependsrec [-A] [pkgname|glob]+\n");
    printf("\twhatrecommends[-A] [pkgname|glob]+\n");
    printf("\twhatsuggests[-A] [pkgname|glob]+\n");
    printf("\twhatprovides [-A] [pkgname|glob]+\n");
    printf("\twhatconflicts [-A] [pkgname|glob]+\n");
    printf("\twhatreplaces [-A] [pkgname|glob]+\n");
    printf("\tverify [pkg|glob]               Verifies the intrgrity of <pkg>, or all packages if omitted by\n");
    printf("\t                                comparing the md5sum of each file with the information stored\n");
    printf("\t                                on the opkg metadata database\n");
    printf("\t-v, --version                   Print version\n");
    printf("\nOptions:\n");
    printf("\t-A                              Query all packages not just those installed\n");
    printf("\t-V[<level>]                     Set verbosity level to <level>.\n");
    printf("\t--verbosity[=<level>]           Verbosity levels:\n");
    printf("\t                                  0 errors only\n");
    printf("\t                                  1 normal messages (default)\n");
    printf("\t                                  2 informative messages\n");
    printf("\t                                  3 debug\n");
    printf("\t                                  4 debug level 2\n");
    printf("\t-f <conf_file>                  Use <conf_file> as the opkg configuration file\n");
    printf("\t--conf <conf_file>\n");
    printf("\t--cache-dir <path>              Specify cache directory.\n");
    printf("\t-t, --tmp-dir <directory>       Specify tmp-dir.\n");
    printf("\t-l, --lists-dir <directory>     Specify lists-dir.\n");
    printf("\t-d <dest_name>                  Use <dest_name> as the the root directory for\n");
    printf("\t--dest <dest_name>              package installation, removal, upgrading.\n");
    printf("\t                                <dest_name> should be a defined dest name from\n");
    printf("\t                                the configuration file, (but can also be a\n");
    printf("\t                                directory name in a pinch).\n");
    printf("\t-o <dir>                        Use <dir> as the root directory for\n");
    printf("\t--offline-root <dir>            offline installation of packages.\n");
    printf("\t--add-dest <name>:<path>        Register destination with given path\n");
    printf("\t--add-arch <arch>:<prio>        Register architecture with given priority\n");
    printf("\t--add-exclude <name>            Register package to be excluded from install\n");
    printf("\t--add-ignore-recommends <name>  Register package to be ignored as a recomendee\n");
    printf("\t--prefer-arch-to-version        Use the architecture priority package rather\n");
    printf("\t                                than the higher version one if more\n");
    printf("\t                                than one candidate is found.\n");
    printf("\t--combine                       Combine upgrade and install operations, this\n");
    printf("\t                                may be needed to resolve dependency issues.\n");
    printf("\t                                Only available for the internal solver backend.\n");
    printf("\t--fields <field1>,<field2>      Limit display information to the specified fields\n");
    printf("\t                                plus the package name. Valid for info and status.\n");
    printf("\t--short-description             Display only the first line of the description.\n");
    printf("\t--size                          Print package size when listing available packages\n");

    printf("\nForce Options:\n");
    printf("\t--force-depends                 Install/remove despite failed dependencies\n");
    printf("\t--force-maintainer              Overwrite preexisting config files\n");
    printf("\t--force-reinstall               Reinstall package(s)\n");
    printf("\t--force-overwrite               Overwrite files from other package(s)\n");
    printf("\t--force-downgrade               Allow opkg to downgrade packages\n");
    printf("\t--force-space                   Disable free space checks\n");
    printf("\t--force-postinstall             Run postinstall scripts even in offline mode\n");
    printf("\t--force-remove                  Remove package even if prerm script fails\n");
    printf("\t--force-checksum                Don't fail on checksum mismatches\n");
    printf("\t--noaction                      No action -- test only\n");
    printf("\t--download-only                 No action -- download only\n");
    printf("\t--nodeps                        Do not follow dependencies\n");
    printf("\t--no-install-recommends         Do not install any recommended packages\n");
    printf("\t--force-removal-of-dependent-packages\n");
    printf("\t                                Remove package and all dependencies\n");
    printf("\t--autoremove                    Remove packages that were installed\n");
    printf("\t                                automatically to satisfy dependencies\n");
    printf("\t--host-cache-dir                Don't place cache in offline root dir.\n");
    printf("\t--volatile-cache                Use volatile cache.\n");
    printf("\t                                Volatile cache will be cleared on exit\n");
    printf("\n");

    printf(" glob could be something like 'pkgname*' '*file*' or similar\n");
    printf(" e.g. opkg info 'libstd*' or opkg search '*libop*' or opkg remove 'libncur*'\n");

    /* --force-removal-of-essential-packages        Let opkg remove essential packages.
     * Using this option is almost guaranteed to break your system, hence this option
     * is not even advertised in the usage statement. */

    exit(1);
}

int main(int argc, char *argv[])
{
    int opts, err = -1;
    char *cmd_name = NULL;
    opkg_cmd_t *cmd;
    int nocheckfordirorfile;
    int noreadfeedsfile;
    int noloadconf;

    if (opkg_conf_init())
        goto err0;

    opkg_config->verbosity = NOTICE;

    opts = args_parse(argc, argv);
    if (opts == argc || opts < 0) {
        fprintf(stderr, "opkg must have one sub-command argument\n");
        usage();
    }

    cmd_name = argv[opts++];

    nocheckfordirorfile = !strcmp(cmd_name, "print-architecture")
        || !strcmp(cmd_name, "print_architecture")
        || !strcmp(cmd_name, "print-installation-architecture")
        || !strcmp(cmd_name, "print_installation_architecture")
        || !strcmp(cmd_name, "compare_versions")
        || !strcmp(cmd_name, "compare-versions")
        || !strcmp(cmd_name, "clean");

    noreadfeedsfile = !strcmp(cmd_name, "flag")
        || !strcmp(cmd_name, "configure")
        || !strcmp(cmd_name, "remove")
        || !strcmp(cmd_name, "files")
        || !strcmp(cmd_name, "search")
        || !strcmp(cmd_name, "list_installed")
        || !strcmp(cmd_name, "list-installed")
        || !strcmp(cmd_name, "list_changed_conffiles")
        || !strcmp(cmd_name, "list-changed-conffiles")
        || !strcmp(cmd_name, "status")
        || !strcmp(cmd_name, "clean");

    noloadconf = !strcmp(cmd_name, "compare_versions")
        || !strcmp(cmd_name, "compare-versions");

    cmd = opkg_cmd_find(cmd_name);
    if (cmd == NULL) {
        fprintf(stderr, "%s: unknown sub-command %s\n", argv[0], cmd_name);
        usage();
    }

    opkg_config->pfm = cmd->pfm;

    if (!noloadconf) {
        if (opkg_conf_load())
            goto err0;
    }

    if (!nocheckfordirorfile) {
        if (!noreadfeedsfile) {
            if (pkg_hash_load_feeds())
                goto err1;
        }

        if (pkg_hash_load_status_files())
            goto err1;
    }

    if (cmd->requires_args && opts == argc) {
        fprintf(stderr, "%s: the ``%s'' command requires at least one argument\n",
                argv[0], cmd_name);
        usage();
    }

    err = opkg_cmd_exec(cmd, argc - opts, (const char **)(argv + opts));

    opkg_download_cleanup();
 err1:
    opkg_conf_deinit();

 err0:
    return err;
}
